File: firebase/functions/lib/events/live_meetings/update_presence_status.dart
import 'dart:async';

import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart'
    hide CloudFunction;
import '../../cloud_function.dart';
import '../../utils/firestore_utils.dart';
import 'package:data_models/firestore/event.dart';
import 'package:data_models/utils.dart';

/// Follow https://firebase.google.com/docs/firestore/solutions/presence
class UpdatePresenceStatus implements CloudFunction {
  @override
  final String functionName = 'UpdatePresenceStatus';

  FutureOr<void> action(
    Change<DataSnapshot> change,
    EventContext context,
  ) async {
    try {
      final realtimeDatabasePresence =
          change.after.val() as Map<String, dynamic>;
      final afterPresenceUpdateTime = DateTime.fromMillisecondsSinceEpoch(
        realtimeDatabasePresence['last_changed'],
      );

      if (realtimeDatabasePresence['state'] != 'offline') return;

      print('change in presence: $realtimeDatabasePresence');

      // If the current timestamp for this data is newer than
      // the data that triggered this event, we exit this function.
      final currentPresenceSnapshot = await change.after.ref.once('value');
      final currentPresenceValue =
          currentPresenceSnapshot.val() as Map<String, dynamic>;
      final currentPresenceTimestamp = DateTime.fromMillisecondsSinceEpoch(
        currentPresenceValue['last_changed'],
      );
      if (currentPresenceTimestamp.isAfter(afterPresenceUpdateTime)) {
        print('presence already changed so ignoring');
        return null;
      }

      print('updating all participants to disconnected');
      await _updateEventStatusesToOffline(
        userId: context.params['uid']!,
        updateTime: afterPresenceUpdateTime,
      );
    } catch (e, stacktrace) {
      print('Error during action $functionName');
      print(e);
      print(stacktrace);
      rethrow;
    }
  }

  Future<void> _updateEventStatusesToOffline({
    required String userId,
    required DateTime updateTime,
  }) async {
    final documentsToUpdate = await firestore
        .collectionGroup('event-participants')
        .where(Participant.kFieldId, isEqualTo: userId)
        .where(Participant.kFieldIsPresent, isEqualTo: true)
        .get();

    await Future.wait(
      documentsToUpdate.documents.map(
        (doc) => firestore.runTransaction((transaction) async {
          final liveParticipantData = await transaction.get(doc.reference);
          final participant = Participant.fromJson(
            firestoreUtils.fromFirestoreJson(liveParticipantData.data.toMap()),
          );
          final lastUpdatedTime = participant.lastUpdatedTime;

          if (lastUpdatedTime != null && lastUpdatedTime.isAfter(updateTime)) {
            print('status already updated so ignoring');
            return;
          }

          transaction.update(
            doc.reference,
            UpdateData.fromMap(
              jsonSubset(
                [
                  Participant.kFieldIsPresent,
                  Participant.kFieldLastUpdatedTime,
                  Participant.kFieldCurrentBreakoutRoomId,
                ],
                firestoreUtils.toFirestoreJson(
                  participant
                      .copyWith(
                        isPresent: false,
                        currentBreakoutRoomId: '',
                      )
                      .toJson()
                    ..[Participant.kFieldLastUpdatedTime] =
                        Timestamp.fromDateTime(updateTime),
                ),
              ),
            ),
          );
        }),
      ),
    );
  }

  @override
  void register(FirebaseFunctions functions) {
    functions[functionName] = functions
        .runWith(
          RuntimeOptions(timeoutSeconds: 60, memory: '1GB', minInstances: 0),
        )
        .database
        .ref('/status/{uid}')
        .onUpdate(action);
  }
}

File: firebase/functions/src/events/live_meetings/update_presence_status.ts
import * as functions from 'firebase-functions';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';
import { CloudFunction } from '../../cloud_function';
import { Participant, ParticipantStatus } from '../../types';

export class UpdatePresenceStatus implements CloudFunction {
  readonly functionName = 'UpdatePresenceStatus';

  async action(change: functions.Change<functions.database.DataSnapshot>, context: functions.EventContext): Promise<void> {
    try {
      const realtimeDatabasePresence = change.after.val() as Record<string, unknown>;
      const afterPresenceUpdateTime = new Date(realtimeDatabasePresence['last_changed'] as number);

      if (realtimeDatabasePresence['state'] !== 'offline') return;

      console.log('change in presence:', realtimeDatabasePresence);

      const currentPresenceSnapshot = await change.after.ref.once('value');
      const currentPresenceValue = currentPresenceSnapshot.val() as Record<string, unknown>;
      const currentPresenceTimestamp = new Date(currentPresenceValue['last_changed'] as number);

      if (currentPresenceTimestamp > afterPresenceUpdateTime) {
        console.log('presence already changed so ignoring');
        return;
      }

      console.log('updating all participants to disconnected');
      await this._updateEventStatusesToOffline({
        userId: context.params['uid']!,
        updateTime: afterPresenceUpdateTime,
      });
    } catch (e) {
      console.error(`Error during action ${this.functionName}`, e);
      throw e;
    }
  }

  private async _updateEventStatusesToOffline({
    userId,
    updateTime,
  }: {
    userId: string;
    updateTime: Date;
  }): Promise<void> {
    const documentsToUpdate = await firestore
      .collectionGroup('event-participants')
      .where('id', '==', userId)
      .where('isPresent', '==', true)
      .get();

    await Promise.all(
      documentsToUpdate.docs.map((doc) =>
        firestore.runTransaction(async (transaction) => {
          const liveParticipantSnap = await transaction.get(doc.ref);
          const participant = firestoreUtils.fromFirestoreJson(liveParticipantSnap.data() ?? {}) as unknown as Participant;
          const lastUpdatedTime = (participant as any).lastUpdatedTime as Date | undefined;

          if (lastUpdatedTime && lastUpdatedTime > updateTime) {
            console.log('status already updated so ignoring');
            return;
          }

          transaction.update(doc.ref, {
            isPresent: false,
            currentBreakoutRoomId: '',
            lastUpdatedTime: updateTime,
          });
        })
      )
    );
  }

  register(_functions: typeof functions): functions.CloudFunction<unknown> {
    return _functions
      .runWith({ timeoutSeconds: 60, memory: '1GB', minInstances: 0 })
      .database.ref('/status/{uid}')
      .onUpdate((change, context) => this.action(change, context));
  }
}
